demo-circuit/src/naive_threshold_sig_w_key_rotation/mod.rs

```diff
@@ -0,0 +1,789 @@
+        //     if (wcert_signatures[i] != NULL) {
+        //         val is_valid_sig = verify_signature(sc_wcert_hash, wcert_signatures[i], signing_keys[i])      // order of signatures should corresponed the order of keys in signing_keys
+        //         require(is_valid_sig == true)
+        //         sigNum = sigNum + 1    
+        //     }
+        // }
+        // require(sigNum > VALIDATORS_SIZE/2)  CR: This is replaced by the threshold value
+   
+        //Alloc the b's as witnesses
+        let mut bs_g = Vec::with_capacity(log_max_pks + 1);
+        for (i, b) in self.b.iter().enumerate() {
+            let b_g = Boolean::alloc(cs.ns(|| format!("alloc b_{}", i)), || {
+                b.ok_or(SynthesisError::Other(format!("AssignmentMissing b {}", i)))
+            })?;
+            bs_g.push(b_g);
+        }
+
+        //Pack the b's into a field element
+        let b_field = FieldElementGadget::from_bits(
+            cs.ns(|| "pack the b's into a field element"),
+            bs_g.as_slice(),
+        )?;
+
+        // 4 ************** Verify quality (We assume `quality` is the number of signatures in `signatures` vector)  *************
+        // require(quality == sigNum)
+
+        //Enforce threshold
+        valid_signatures
+            .sub(cs.ns(|| "valid_signatures - threshold"), &threshold_g)?
+            .enforce_equal(cs.ns(|| "threshold check"), &b_field)?;
+
+        // **************** 5 Verify signatures of updated validators keys ****************
+        // for (i <- 0 to (VALIDATORS_SIZE-1)) {
+        //
+        //     if (updated_signing_keys[i] != signing_keys[i]) {
+        //         val msg_to_sign = updated_signing_keys[i];
+        //         val sk_sig = updated_signing_keys_sk_signatures[i];
+        //         val mk_sig = updated_signing_keys_mk_signatures[i];
+        //         val is_valid_sk_sig = verify_signature(msg_to_sign, sk_sig, signing_keys[i])      // updated signing key should be signed old signing key
+        //         val is_valid_mk_sig = verify_signature(msg_to_sign, mk_sig, master_keys[i])       // updated signing key should be signed old master key
+        //         require(is_valid_sk_sig == is_valid_mk_sig == true)
+        //     }
+        //
+        //     if (updated_master_keys[i] != master_keys[i]) {
+        //         val msg_to_sign = updated_master_keys[i];
+        //         val sk_sig = updated_master_keys_sk_signatures[i];
+        //         val mk_sig = updated_master_keys_mk_signatures[i];
+        //         val is_valid_sk_sig = verify_signature(msg_to_sign, sk_sig, signing_keys[i])      // updated master key should be signed old signing key
+        //         val is_valid_mk_sig = verify_signature(msg_to_sign, mk_sig, master_keys[i])       // updated master key should be signed old master key
+        //         require(is_valid_sk_sig == is_valid_mk_sig == true)
+        //     }
+        // }
```

Comment:
Let's document the code step by step if possible, rather than putting a big comment at the top of each section. It will really help readability